stream_alert/classifier/parsers.py

```diff
@@ -0,0 +1,808 @@
+
+        return True
+
+    def _validate_schema(self):
+        """Ensure the schema, including optional keys, etc, is defined properly
+
+        All optional top level keys should also be defined within the schema itself,
+        and just included in the optional top level keys list. The same applies to
+        optional envelope keys and the envelope keys/schema.
+
+        In the future, this function can be added to so that it also performs validation
+        of types defined for values and so forth.
+
+        Returns:
+            bool: True if this schema is valid, False otherwise
+        """
+        # TODO (ryandeivert): check the value of types defined in the schema to ensure
+        # they are valid before erroring out at the _convert_type stage
+        values = [
+            (self.schema, self.optional_top_level_keys),
+            (self.envelope_schema, self.optional_envelope_keys)
+        ]
+
+        return all(optionals.issubset(schema) for schema, optionals in values)
+
+    def _add_parse_result(self, record, valid, envelope):
+        """Add the result of parsing to the proper array, updating the envelope in the process
+
+        Args:
+            record (dict): Value of the parsed (or unparsed, if failed) record
+            valid (bool): True if the parsing was successful, False otherwise
+            envelope (dict): Extracted envelope to be added to the parsed record
+        """
+        if not valid:
+            self._invalid_parses.append(record)
+            return
+
+        self._apply_envelope(record, envelope)
+        self._add_optional_keys(record, self.schema, self.optional_top_level_keys)
+
+        self._valid_parses.append(record)
+
+    def _extract_envelope(self, payload):
+        """Extract envelope key/values from the original payload
+
+        Args:
+            payload (dict): The original record to pull envelope keys from
+
+        Returns:
+            dict: Key/values extracted from the log to be used as the envelope
+        """
+        if not self.envelope_schema:
+            return
+
+        LOGGER.debug('Extracting envelope keys')
+        return {
+            key: payload[key]
+            for key in self.envelope_schema
+            if key in payload  # This is fine since some of these may be optional
+        }
+
+    def _json_path_records(self, payload):
+        """Extract records from the original json payload using a provided JSON path
+
+        Args:
+            payload (dict): The parsed json data
+
+        Returns:
+            list: A list of JSON records extracted via JSON path
+        """
+        # Handle jsonpath extraction of records
+        LOGGER.debug('Parsing records with JSONPath: %s', self.json_path)
+
+        return jmespath.search(self.json_path, payload)
+
+    def parse(self, data):
+        """Main parser method to be handle parsing of the passed data
+
+        Args:
+            data (str|dict): Data to be parsed.
```

Comment:
when would this be a dict and not a string?

Reply:
nm i see it calls itself now

Reply:
i take it back, i'm curious again

Reply:
spoke offline a bit